geopmdpy/geopmdpy/session.py

```diff
@@ -156,25 +156,167 @@ def write_report(self, report, do_gather=True):
+      - header_names: Add custom trace column headers.
+      - trace_out: Provide custom trace output values.
+
+    Example usage:
+        class MyAgent(Agent):
+            def update_parser(self, parser):
+                parser.add_argument('--foo', ...)
+                return parser
+            def update_args(self, args):
+                self._foo = args.foo
+                return args
+            def update_loop(self):
+                # Custom logic here
+                pass
+
+        main(agent=MyAgent())
+
+    """
+    def __init__(self):
+        """Initialize the agent. Override to set up agent state."""
+        pass
+
+    def update_parser(self, parser):
+        """Update the argument parser with agent-specific options.
+
+        Args:
+            parser (argparse.ArgumentParser): The parser to update.
+
+        Returns:
+            argparse.ArgumentParser: The updated parser.
+
+        """
+        return parser
+
+    def update_args(self, args):
+        """Process parsed arguments and update agent state.
+
+        Args:
+            args (argparse.Namespace): Parsed command-line arguments.
+
+        Returns:
+            argparse.Namespace: The (possibly updated) arguments.
+
+        """
+        return args
+
+    def signal_config_override(self):
+        """Override the default signal configuration.
+
+        Returns:
+            str or None: A string containing the signal configuration,
+                         or None to use the Session default (stdin).
+
+        """
+        return None
+
+    def run_begin(self):
+        """Called by Session at the start of each run.
+
+        Override to perform setup before the session starts. Calls to
+        pio.push_signal() and pio.push_control() may be made here.
+
+        """
+        pass
+
+    def run_end(self):
+        """Called by Session at the end of each run.
+
+        Override to perform cleanup after the session ends.  Releases any
+        resources allocated in the run_begin() method.
+
+        """
+        pass
+
+    def update_loop(self):
+        """Called periodically by Session during the run.
+
+        Override to implement periodic agent logic. This method is called by the
+        Session implementation just after a call to pio.read_batch() in the
+        timed loop, so the Agent implementation should not call
+        pio.read_batch().  Calls to pio.sample(), pio.adjust() and
+        pio.write_batch() may be part of the update_loop() implementation.
+
+        """
+        pass
+
+    def header_names(self):
+        """Return additional trace column headers provided by the agent.
 
-    This object's run() method is the main entry point for
-    geopmsession command line tool.  The inputs to run() are derived
-    from the command line options provided by the user.
+        Returns:
+            list of str: List of header names for custom trace columns.
+
+        """
+        return []
+
+    def trace_out(self):
```

Comment:
Can a check be added to ensure the number of outputs for trace matches the header_names?